### ISSUE TITLE
Recursive function call is not a closure call if the function is closure because it's used as variable

### ISSUE THREAD

**Issue by rhysd** [rhysd/gocaml#11]
Repro:

```ml
let rec sum i = if i = 1 then 1 else i + (sum (i - 1)) in
let rec get_sum _ = sum in
println_int ((get_sum ()) 42)
```

Error:

```
: Incorrect number of arguments passed to called function!
  %0 = call i64 @"sum$t1"(i64 %"i$t2")
```

The `sum` is transformed as known function at `e1` in `let rec sum i = e1 in e2`, but when transforming `e2`, it is fallbacked into closure because it is used as variable in `e2`. But the recursive function call remains as direct function application.

**Comment by rhysd** [comment]
I additionally found the case where `e2` may contain wrong direct function call.

```ml
let rec sum i = if i = 1 then 1 else i + (sum (i - 1)) in
let x = sum 20 in
let rec get_sum _ = sum in
println_int ((get_sum ()) 10)
```

The `sum 20` function call is converted to direct `app` instruction but actually it should be closure call because `sum` is used as a variable after.

**Comment by rhysd** [comment]
There are at least 2 ways to solve this problem:

1. Do not see `CLOSURE_CALL` at visiting `app` insn at codegen. We can know the callee of `app` insn is a closure or not by function table, external symbols table and closures table. `CLOSURE_CALL` might be set by additional traverse after closure transform and before codegen. (Anyway, we need to arrange basic blocks in GCIL program, maybe...). In this case, we cannot know the closure is a recursive function so we should always emit the closure object for itself in its body.
2. When we've found that the function should be closure because it is used as variable, then retry closure transform for `e1` and `e2` in `let rec f x = e1 in e2`.

**Comment by rhysd** [comment]
I chose 1. and set `CLOSURE_CALL` properly at post process of closure transform (`IsRecursive` is also set).

**Comment by rhysd** [comment]
Fixed in 40738e6f1c03c777f56a7ed6c651ae579136f393 and some commits after it
